File: internal/sstable/writer.go
package sstable

import (
	"encoding/binary"
	"fmt"
	"hash/crc32"
	"io"
	"os"
	"sort"

	"bigtable/internal/bloom"
	"bigtable/internal/record"
)

func WriteSnapshot(path string, snapshot map[string]record.Record, sparseGap int) error {
	if sparseGap <= 0 {
		sparseGap = 10
	}

	tmpPath := path + ".tmp"

	f, err := os.Create(tmpPath)
	if err != nil {
		return fmt.Errorf("create sstable tmp: %w", err)
	}

	keys := make([]string, 0, len(snapshot))
	for k := range snapshot {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	indexEntries := make([]SparseIndexEntry, 0, len(keys))

	// Write every record exactly as it exists in the snapshot.
	// Tombstones are first-class records and must be persisted too.
	for i, k := range keys {
		rec := snapshot[k]

		offset, err := f.Seek(0, io.SeekCurrent)
		if err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("get write offset: %w", err)
		}

		if i%sparseGap == 0 {
			keyCopy := append([]byte(nil), rec.Key...)
			indexEntries = append(indexEntries, SparseIndexEntry{
				Key:    keyCopy,
				Offset: uint64(offset),
			})
		}

		payload, err := rec.Encode()
		if err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("encode record: %w", err)
		}

		var lenBuf [4]byte
		binary.BigEndian.PutUint32(lenBuf[:], uint32(len(payload))|checksumFlag)

		var crcBuf [4]byte
		binary.BigEndian.PutUint32(crcBuf[:], crc32.ChecksumIEEE(payload))

		if err := writeFull(f, lenBuf[:]); err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("write record length: %w", err)
		}
		if err := writeFull(f, payload); err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("write record payload: %w", err)
		}
		if err := writeFull(f, crcBuf[:]); err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("write record crc: %w", err)
		}
	}

	bf := bloom.NewForKeys(len(keys))
	for _, k := range keys {
		bf.Add([]byte(k))
	}

	bloomOffset, err := f.Seek(0, io.SeekCurrent)
	if err != nil {
		_ = f.Close()
		_ = os.Remove(tmpPath)
		return fmt.Errorf("get bloom offset: %w", err)
	}

	bloomBytes := bf.MarshalBinary()
	if err := writeFull(f, bloomBytes); err != nil {
		_ = f.Close()
		_ = os.Remove(tmpPath)
		return fmt.Errorf("write bloom filter: %w", err)
	}

	indexOffset, err := f.Seek(0, io.SeekCurrent)
	if err != nil {
		_ = f.Close()
		_ = os.Remove(tmpPath)
		return fmt.Errorf("get index offset: %w", err)
	}

	for _, entry := range indexEntries {
		var keyLenBuf [4]byte
		binary.BigEndian.PutUint32(keyLenBuf[:], uint32(len(entry.Key)))

		var offsetBuf [8]byte
		binary.BigEndian.PutUint64(offsetBuf[:], entry.Offset)

		if err := writeFull(f, keyLenBuf[:]); err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("write index key len: %w", err)
		}
		if err := writeFull(f, entry.Key); err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("write index key: %w", err)
		}
		if err := writeFull(f, offsetBuf[:]); err != nil {
			_ = f.Close()
			_ = os.Remove(tmpPath)
			return fmt.Errorf("write index offset: %w", err)
		}
	}

	footer := make([]byte, newFooterSize)
	binary.BigEndian.PutUint64(footer[0:8], uint64(indexOffset))
	binary.BigEndian.PutUint32(footer[8:12], uint32(len(indexEntries)))
	binary.BigEndian.PutUint32(footer[12:16], uint32(len(keys)))
	binary.BigEndian.PutUint32(footer[16:20], uint32(sparseGap))
	binary.BigEndian.PutUint64(footer[20:28], uint64(bloomOffset))
	binary.BigEndian.PutUint32(footer[28:32], uint32(len(bloomBytes)))
	copy(footer[32:40], []byte(newMagicValue))

	if err := writeFull(f, footer); err != nil {
		_ = f.Close()
		_ = os.Remove(tmpPath)
		return fmt.Errorf("write footer: %w", err)
	}

	if err := f.Sync(); err != nil {
		_ = f.Close()
		_ = os.Remove(tmpPath)
		return fmt.Errorf("sync sstable: %w", err)
	}

	if err := f.Close(); err != nil {
		_ = os.Remove(tmpPath)
		return fmt.Errorf("close sstable tmp: %w", err)
	}

	if err := os.Rename(tmpPath, path); err != nil {
		_ = os.Remove(tmpPath)
		return fmt.Errorf("rename sstable tmp: %w", err)
	}

	return nil
}

func writeFull(f *os.File, data []byte) error {
	for len(data) > 0 {
		n, err := f.Write(data)
		if err != nil {
			return err
		}
		data = data[n:]
	}
	return nil
}

File: internal/sstable/writer.cpp
#include "writer.h"
#include <fstream>
#include <filesystem>
#include <algorithm>
#include <vector>
#include <stdexcept>
#include <cstring>
#include <iostream>

namespace fs = std::filesystem;

namespace bigdb::sstable {

namespace {

uint32_t crc32_ieee(const std::vector<uint8_t>& data) {
    uint32_t crc = 0xFFFFFFFF;
    for (uint8_t byte : data) {
        crc ^= byte;
        for (int i = 0; i < 8; i++) {
            uint32_t mask = -(crc & 1);
            crc = (crc >> 1) ^ (0xEDB88320 & mask);
        }
    }
    return ~crc;
}

void writeFull(std::ofstream& f, const std::vector<uint8_t>& data) {
    f.write(reinterpret_cast<const char*>(data.data()), data.size());
    if (f.fail()) {
        throw std::runtime_error("write failed");
    }
}

} // namespace

void WriteSnapshot(const std::string& path, const std::map<std::string, record::Record>& snapshot, int sparse_gap) {
    if (sparse_gap <= 0) {
        sparse_gap = 10;
    }

    std::string tmpPath = path + ".tmp";
    std::ofstream f(tmpPath, std::ios::binary | std::ios::trunc);
    if (!f) {
        throw std::runtime_error("create sstable tmp failed");
    }

    std::vector<std::string> keys;
    keys.reserve(snapshot.size());
    for (const auto& [k, _] : snapshot) {
        keys.push_back(k);
    }
    // std::map is already sorted, but let's be sure.
    std::sort(keys.begin(), keys.end());

    std::vector<SparseIndexEntry> indexEntries;

    for (size_t i = 0; i < keys.size(); i++) {
        const auto& k = keys[i];
        const auto& rec = snapshot.at(k);

        uint64_t offset = f.tellp();

        if (i % sparse_gap == 0) {
            indexEntries.push_back({rec.key, offset});
        }

        std::vector<uint8_t> payload = rec.encode();

        std::vector<uint8_t> lenBuf(4);
        uint32_t lenWithFlag = static_cast<uint32_t>(payload.size()) | ChecksumFlag;
        lenBuf[0] = (lenWithFlag >> 24) & 0xFF;
        lenBuf[1] = (lenWithFlag >> 16) & 0xFF;
        lenBuf[2] = (lenWithFlag >> 8) & 0xFF;
        lenBuf[3] = lenWithFlag & 0xFF;

        std::vector<uint8_t> crcBuf(4);
        uint32_t crc = crc32_ieee(payload);
        crcBuf[0] = (crc >> 24) & 0xFF;
        crcBuf[1] = (crc >> 16) & 0xFF;
        crcBuf[2] = (crc >> 8) & 0xFF;
        crcBuf[3] = crc & 0xFF;

        writeFull(f, lenBuf);
        writeFull(f, payload);
        writeFull(f, crcBuf);
    }

    auto bf = bloom::Filter::NewForKeys(keys.size());
    for (const auto& k : keys) {
        std::vector<uint8_t> keyBytes(k.begin(), k.end());
        bf->Add(keyBytes);
    }

    uint64_t bloomOffset = f.tellp();
    std::vector<uint8_t> bloomBytes = bf->MarshalBinary();
    writeFull(f, bloomBytes);

    uint64_t indexOffset = f.tellp();

    for (const auto& entry : indexEntries) {
        std::vector<uint8_t> keyLenBuf(4);
        uint32_t keyLen = entry.key.size();
        keyLenBuf[0] = (keyLen >> 24) & 0xFF;
        keyLenBuf[1] = (keyLen >> 16) & 0xFF;
        keyLenBuf[2] = (keyLen >> 8) & 0xFF;
        keyLenBuf[3] = keyLen & 0xFF;

        std::vector<uint8_t> offsetBuf(8);
        offsetBuf[0] = (entry.offset >> 56) & 0xFF;
        offsetBuf[1] = (entry.offset >> 48) & 0xFF;
        offsetBuf[2] = (entry.offset >> 40) & 0xFF;
        offsetBuf[3] = (entry.offset >> 32) & 0xFF;
        offsetBuf[4] = (entry.offset >> 24) & 0xFF;
        offsetBuf[5] = (entry.offset >> 16) & 0xFF;
        offsetBuf[6] = (entry.offset >> 8) & 0xFF;
        offsetBuf[7] = entry.offset & 0xFF;

        writeFull(f, keyLenBuf);
        writeFull(f, entry.key);
        writeFull(f, offsetBuf);
    }

    std::vector<uint8_t> footer(NewFooterSize, 0);
    // index_offset
    footer[0] = (indexOffset >> 56) & 0xFF; footer[1] = (indexOffset >> 48) & 0xFF;
    footer[2] = (indexOffset >> 40) & 0xFF; footer[3] = (indexOffset >> 32) & 0xFF;
    footer[4] = (indexOffset >> 24) & 0xFF; footer[5] = (indexOffset >> 16) & 0xFF;
    footer[6] = (indexOffset >> 8) & 0xFF;  footer[7] = indexOffset & 0xFF;
    
    // index_count
    uint32_t indexCount = indexEntries.size();
    footer[8] = (indexCount >> 24) & 0xFF; footer[9] = (indexCount >> 16) & 0xFF;
    footer[10] = (indexCount >> 8) & 0xFF; footer[11] = indexCount & 0xFF;
    
    // record_count
    uint32_t recordCount = keys.size();
    footer[12] = (recordCount >> 24) & 0xFF; footer[13] = (recordCount >> 16) & 0xFF;
    footer[14] = (recordCount >> 8) & 0xFF;  footer[15] = recordCount & 0xFF;
    
    // sparse_gap
    footer[16] = (sparse_gap >> 24) & 0xFF; footer[17] = (sparse_gap >> 16) & 0xFF;
    footer[18] = (sparse_gap >> 8) & 0xFF;  footer[19] = sparse_gap & 0xFF;
    
    // bloom_offset
    footer[20] = (bloomOffset >> 56) & 0xFF; footer[21] = (bloomOffset >> 48) & 0xFF;
    footer[22] = (bloomOffset >> 40) & 0xFF; footer[23] = (bloomOffset >> 32) & 0xFF;
    footer[24] = (bloomOffset >> 24) & 0xFF; footer[25] = (bloomOffset >> 16) & 0xFF;
    footer[26] = (bloomOffset >> 8) & 0xFF;  footer[27] = bloomOffset & 0xFF;
    
    // bloom_size
    uint32_t bloomSize = bloomBytes.size();
    footer[28] = (bloomSize >> 24) & 0xFF; footer[29] = (bloomSize >> 16) & 0xFF;
    footer[30] = (bloomSize >> 8) & 0xFF;  footer[31] = bloomSize & 0xFF;
    
    // magic
    std::memcpy(footer.data() + 32, NewMagicValue, std::strlen(NewMagicValue));

    writeFull(f, footer);
    f.flush();
    f.close();

    fs::rename(tmpPath, path);
}

} // namespace bigdb::sstable
